---
title: "scRNAseq PBMC Oelen"
Author: "Pooja"
date: "2025"
output:
  html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval = TRUE,
                      message = FALSE,
                      warning = FALSE)
options(max.print=200)

```
Load necessary libraries
```{r}
suppressPackageStartupMessages({
library(Seurat)
library(data.table)
library(dplyr)
library(ggplot2)
library(dittoSeq)
  library(speckle)
  library(patchwork)
  library(ComplexHeatmap)
  library(tidyr)
  library(colorRamp2)
  library(readxl)
  library(scCustomize)
  library(reshape2)
  library(knitr)
library(tibble)
})
```
INPUT FILES

Rename the barcode, matrix and features file to barcodes.tsv.gz, features.tsv.gz, matrix.tsv.gz before running this step
```{r, eval= FALSE}
# Read the 10X files
pbmc_v2_data <- Read10X("/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/n_qc_v2_data", gene.column = 1, cell.column = 1)
#pbmc_v3_data <-  Read10X("/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/n_qc_v3_data", gene.column = 1, cell.column = 1)
# Create the seurat object
pbmc_v2 <- CreateSeuratObject(counts = pbmc_v2_data, project = "pbmc_v2", min.cells = 3, min.features = 200)
#pbmc_v3 <- CreateSeuratObject(counts = pbmc_v3_data, project = "pbmc_v3", min.cells = 3, min.features = 200)
# Save the file for further use
#saveRDS(pbmc, file = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/pbmc.rds")
```
Read other meta data files
```{r}

# Read the saved file
#pbmc <- readRDS("/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/pbmc.rds")

# Read metadata


# Omit rows with NA values
#pathogen.info <- na.omit(pathogen.info)
```
Add the info
```{r, eval = FALSE}
# Add cell type info and pathogen info
pbmc_v2$cell_type <- cell.info$cell_type_lowerres[match(colnames(pbmc_v2), cell.info$barcode)]
pbmc_v2$cell_subtype <- cell.info$cell_type[match(colnames(pbmc_v2), cell.info$barcode)]
pbmc_v2$subject <- pathogen.info$assignment[match(colnames(pbmc_v2), pathogen.info$barcode)]
pbmc_v2$timepoint <- pathogen.info$timepoint[match(colnames(pbmc_v2), pathogen.info$barcode)]

# Add cell type info and pathogen info
#pbmc_v3$cell_type <- cell.info$cell_type_lowerres[match(colnames(pbmc_v3), cell.info$barcode)]
#pbmc_v3$cell_subtype <- cell.info$cell_type[match(colnames(pbmc_v3), cell.info$barcode)]
#pbmc_v3$subject <- pathogen.info$assignment[match(colnames(pbmc_v3), pathogen.info$barcode)]
#pbmc_v3$timepoint <- pathogen.info$timepoint[match(colnames(pbmc_v3), pathogen.info$barcode)]


```
```{r, eval = FALSE}
combined <- merge(
  x = pbmc_v2,
  y = pbmc_v3,
  add.cell.ids = c(
    "pbmc_v2", "pbmc_v3")
)
combined_so <- JoinLayers(combined)
saveRDS(combined_so, file ="/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/v2_v3_pbmc.rds")
```
```{r, eval = FALSE}
combined_so <- readRDS("/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/v2_v3_pbmc.rds")

# there is an NA in one of the subjects and timepoint, remove it to proceed
met.data <- combined_so@meta.data$subject
which(is.na(pbmc_v2@meta.data$subject), arr.ind = TRUE)
#800294
cell_to_remove <- "pbmc_v3_TTGTTTGCATTCTGTT_190122_lane2"
combined_so <- subset(combined_so, cells = setdiff(Cells(combined_so), cell_to_remove))

# check if it is successfuly removed

which(is.na(combined_so@meta.data$subject), arr.ind = TRUE)
#saveRDS(combined_so, file ="/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/v2_v3_pbmc.rds")

```
Data Normalisation and scaling
```{r, eval = FALSE}
# Subset untreated samples
pbmc.UT <- subset(pbmc_v2, subset = timepoint == "UT")

#Quality Control
pbmc.UT[["percent.mt"]] <- PercentageFeatureSet(pbmc.UT, pattern = "^MT-")

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
pbmc.UT <- PercentageFeatureSet(pbmc.UT, "^HB[^(P)]", col.name = "percent_hb")

# Visualize QC metrics as a violin plot
VlnPlot(pbmc.UT, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_hb"), ncol = 4)

#Filtering from contaminated data
pbmc.UT <- subset(pbmc.UT, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 8 & percent_hb < 2)

#Normalizing data
pbmc.UT <- NormalizeData(pbmc.UT, normalization.method = "LogNormalize", scale.factor = 10000)

#Identifying highly variable featured
pbmc.UT <- FindVariableFeatures(pbmc.UT, selection.method = "vst", nfeatures = 2000)

all.genes <- rownames(pbmc.UT)
#Scaling the data
pbmc.UT <- ScaleData(pbmc.UT, features = all.genes)
saveRDS(pbmc.UT, file ="/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/pbmc_UT.rds")
```
```{r, eval = FALSE}
#combined_so <- readRDS("/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/v2_v3_pbmc.rds")
#Quality Control

combined_so[["percent.mt"]] <- PercentageFeatureSet(combined_so, pattern = "^MT-")

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
combined_so <- PercentageFeatureSet(combined_so, "^HB[^(P)]", col.name = "percent_hb")

# Visualize QC metrics as a violin plot
#v.plot <- VlnPlot(combined_so, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_hb"), ncol = 4)

#ggsave(filename = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/violin_plot.png", plot = v.plot)
#Filtering from contaminated data
combined_so <- subset(combined_so, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 8 & percent_hb < 2)

#v2.plot <- VlnPlot(combined_so, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent_hb"), ncol = 4)

#ggsave(filename = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/violin_plot2.png", plot = v2.plot)


```

```{r, eval = FALSE}

#Normalizing data
print("Normalising")
combined_so <- NormalizeData(combined_so, normalization.method = "LogNormalize", scale.factor = 10000)
print("Normalising done")

#Identifying highly variable featured
print("Finding var features")
combined_so <- FindVariableFeatures(combined_so, selection.method = "vst", nfeatures = 2000)
print("Found var features")
all.genes <- rownames(combined_so)
#Scaling the data
combined_so <- ScaleData(combined_so, features = all.genes)
print("SCaling complete")
saveRDS(combined_so, file ="/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/v2_v3_pbmc.rds")
```


```{r, eval = FALSE}

table(Idents(combined_so))

combined_so <- SetIdent(combined_so, value = "subject")

cluster.averages <-AverageExpression(combined_so, features = c(Xgenes, Ygenes), return.seurat = TRUE)
saveRDS(cluster.averages, file ="/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/cluster_avg.rds")
```
```{r, eval = FALSE}
cluster.averages <- readRDS("/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/cluster_avg.rds")
Xgenes <- c("ARHGAP4", "STS", "ARSD", "ARSL", "AVPR2", "BRS3", "S100G",
                "CHM", "CLCN4", "DDX3X", "EIF1AX", "EIF2S3", "GPM6B",
                "GRPR", "HCFC1", "L1CAM", "MAOA", "MYCLP1", "NAP1L3",
                "GPR143", "CDK16", "PLXNB3", "PRKX", "RBBP7", "RENBP",
                "RPS4X", "TRAPPC2", "SH3BGRL", "TBL1X", "UBA1", "KDM6A",
                "XG", "XIST", "ZFX", "PUDP", "PNPLA4", "USP9X", "KDM5C",
                "SMC1A", "NAA10", "OFD1", "IKBKG", "PIR", "INE2", "INE1",
                "AP1S2", "GYG2", "MED14", "RAB9A", "ITM2A", "MORF4L2",
                "CA5B", "SRPX2", "GEMIN8", "CTPS2", "CLTRN", "NLGN4X",
                "DUSP21", "ALG13", "SYAP1", "SYTL4", "FUNDC1", "GAB3",
                "RIBC1", "FAM9C", "CA5BP1")

Ygenes <- c("AMELY", "DAZ1", "PRKY", "RBMY1A1", "RBMY1HP", "RPS4Y1", "SRY",
                "TSPY1", "UTY", "ZFY", "KDM5D", "USP9Y", "DDX3Y", "PRY", "XKRY",
                "BPY2", "VCY", "CDY1", "EIF1AY", "TMSB4Y", "CDY2A", "NLGN4Y",
                "PCDH11Y", "HSFY1", "TGIF2LY", "TBL1Y", "RPS4Y2", "HSFY2",
                "CDY2B", "TXLNGY", "CDY1B", "DAZ3", "DAZ2", "DAZ4")


gene.heatmap <- DoHeatmap(cluster.averages, features = c(Xgenes, Ygenes), size = 2, draw.lines = FALSE)

ggsave(filename = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/gene_heatmap2.png", plot = gene.heatmap)
gene.heatmap.x <- DoHeatmap(cluster.averages, features = Xgenes, size = 2, draw.lines = FALSE)
gene.heatmap.y <- DoHeatmap(cluster.averages, features = Ygenes, size = 2, draw.lines = FALSE)


```

```{r, eval = FALSE}

pbmc_v2[["percent.mt"]] <- PercentageFeatureSet(pbmc_v2, pattern = "^MT-")

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
pbmc_v2 <- PercentageFeatureSet(pbmc_v2, "^HB[^(P)]", col.name = "percent_hb")
print("Feature")
#Filtering from contaminated data
pbmc_v2 <- subset(pbmc_v2, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 8 & percent_hb < 2)
print("Subset")
pbmc_v2 <- NormalizeData(pbmc_v2, normalization.method = "LogNormalize", scale.factor = 10000)
print("Norm")
#Identifying highly variable featured
pbmc_v2 <- FindVariableFeatures(pbmc_v2, selection.method = "vst", nfeatures = 2000)
print("Var")
all.genes <- rownames(pbmc_v2)
pbmc_v2 <- ScaleData(pbmc_v2, features = all.genes)
print("Scaled")
saveRDS(pbmc_v2, file ="/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/pbmc_v2.rds")
print("Saved")

```
```{r, eval = FALSE}

pbmc_v3[["percent.mt"]] <- PercentageFeatureSet(pbmc_v3, pattern = "^MT-")

# Percentage hemoglobin genes - includes all genes starting with HB except HBP.
pbmc_v3 <- PercentageFeatureSet(pbmc_v3, "^HB[^(P)]", col.name = "percent_hb")

#Filtering from contaminated data
pbmc_v3 <- subset(pbmc_v3, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 8 & percent_hb < 2)

pbmc_v3 <- NormalizeData(pbmc_v3, normalization.method = "LogNormalize", scale.factor = 10000)

#Identifying highly variable featured
pbmc_v3 <- FindVariableFeatures(pbmc_v3, selection.method = "vst", nfeatures = 2000)

all.genes <- rownames(pbmc_v3)
pbmc_v3 <- ScaleData(pbmc_v3, features = all.genes)
#saveRDS(pbmc_v3, file ="/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/pbmc_v3.rds")
```
```{r, eval = FALSE}
#pbmc <- UpdateSeuratObject(pbmc)

#Idents(pbmc) <- "Subject"

table(Idents(pbmc_v2))
pbmc_v2 <- SetIdent(pbmc_v2, value = "subject")

table(Idents(pbmc_v3))
pbmc_v3 <- SetIdent(pbmc_v3, value = "subject")

cluster.averages.v2 <-AverageExpression(pbmc_v2, return.seurat = TRUE)
cluster.averages.v3 <-AverageExpression(pbmc_v3, return.seurat = TRUE)

```

```{r, eval = FALSE}
gene.heatmap.v2 <- DoHeatmap(cluster.averages.v2, features = c("UTY","DDX3Y","KDM5D","RPS4Y1", "RPS4X", "ZFX", "XIST"), size = 2, draw.lines = FALSE)
gene.heatmap.v3 <- DoHeatmap(cluster.averages.v3, features = c("UTY","DDX3Y","KDM5D","RPS4Y1", "RPS4X", "ZFX", "XIST"), size = 2, draw.lines = FALSE)
ggsave(filename = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/gene_heatmap_v2.png", plot = gene.heatmap.v2)
ggsave(filename = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/normalised_qc/gene_heatmap_v3.png", plot = gene.heatmap.v3)
```
```{r, eval = FALSE}
# Sort the Seurat object by cluster identities
cluster.averages.v2 <- cluster.averages.v2[, order(Idents(cluster.averages.v2))]
cluster.averages.v3 <- cluster.averages.v3[, order(Idents(cluster.averages.v3))]

# Create the heatmap with the sorted order
gene.heatmap.v2 <- DoHeatmap(cluster.averages.v2[, order(Idents(cluster.averages.v2))], features = c("UTY", "DDX3Y", "KDM5D", "RPS4Y1", "RPS4X", "ZFX", "XIST"), size = 2, draw.lines = FALSE)
gene.heatmap.v3 <- DoHeatmap(cluster.averages.v3, features = c("UTY", "DDX3Y", "KDM5D", "RPS4Y1", "RPS4X", "ZFX", "XIST"), size = 2, draw.lines = FALSE)
```

Assign Sex information
```{r, eval = FALSE}

#pbmc.UT <- subset(pbmc_v2, subset = timepoint == "UT")
pbmc.UT <- readRDS("/home/inf-21-2023/thesis/scRNAseq_data_oelen/pbmc.UT.rds")
#Finding and Assigning Sex
table(Idents(pbmc.UT))

pbmc.UT<- SetIdent(pbmc.UT, value = "subject")

#pbmc.UT[["RNA"]] <- as(object = pbmc.UT[["RNA"]], Class = "Assay5")
cluster.averages <- AverageExpression(pbmc.UT, return.seurat = TRUE)

#Assign respective subjects as males females as seen in heatmap
new.Sex.ids <- "Male"
Male<-subset(pbmc.UT, idents = c("2", "12","3","8","16","13","14","22","21","18","23","27","26", "24", "30", "32", "37","44", "45", "41", "39", "58", "60", "64", "65","68", "67", "73", "79", "84","77","78","82","92","91","90"))
Female <-subset(pbmc.UT, idents = c("2", "12","3","8","16","13","14","22","21","18","23","27","26", "24", "30","32", "37","44", "45", "41", "39", "58", "60", "64", "65","68", "67", "73", "79", "84","77","78","82","92","91","90"), invert = TRUE)

#Adding Sex to pbmc.UT metadata
new.Sex.ids <- "Male"
new.Sex.ids <- replicate(77785 , new.Sex.ids)

#to replace n, check length of male metadata
Male$Sex <- new.Sex.ids
Barcode <- rownames(Male@meta.data)
Male$Barcode <- Barcode
# Create a vector to store the Sex information
sex_vector <- rep("Female", ncol(pbmc.UT)) # Initialize with "Female"

# Match and assign "Male" where barcodes match
matched_indices <- match(colnames(pbmc.UT), Male$Barcode)
sex_vector[!is.na(matched_indices)] <- Male$Sex[na.omit(matched_indices)]

# Assign the corrected vector to pbmc.UT metadata
pbmc.UT@meta.data[["Sex"]] <- sex_vector
```

```{r, echo = FALSE, eval = FALSE}
#Check if assignationcorrect
#cluster.averages.g <- AverageExpression(pbmc.UT, group.by =  c("subject", "Sex"), return.seurat = TRUE)
# Add nFeature_RNA from pbmc.UT to the cluster.averages.g metadata
#cluster.averages.g@meta.data$nFeature_RNA <- pbmc.UT@meta.data$nFeature_RNA
#cluster.averages.g$Sex <- colnames(cluster.averages.g)
#cluster.averages.1 <- separate(cluster.averages.g@meta.data, col = "Sex", into = c("subject", "Sex"), sep = "\\_")
#cluster.averages.g$Sex <- cluster.averages.1$Sex[match(cluster.averages.g$nFeature_RNA, cluster.averages.1$nFeature_RNA)]
#DoHeatmap(cluster.averages.g, features = c("UTY","DDX3Y","KDM5D","RPS4Y1", "RPS4X", "EIFAX", "XIST"), group.by ="Sex", size = 2, draw.lines = FALSE)

```
Dimensionality reduction (PCA and UMAP)
```{r, eval = FALSE}
#Linear dimention reduction
pbmc.UT<- RunPCA(pbmc.UT, features = VariableFeatures(object = pbmc.UT))
#ElbowPlot(pbmc.UT)
#Elbow plot shows the first 15 components contirbuting to variation
#Cluster cells
pbmc.UT <- FindNeighbors(pbmc.UT, dims = 1:15)
pbmc.UT<- FindClusters(pbmc.UT, resolution = c(0.1,0.3, 0.5, 0.7, 1))
#DimPlot(pbmc.UT, group.by = "RNA_snn_res.0.1", label = TRUE)

#Non linear dimentional reduction
pbmc.UT <- RunUMAP(pbmc.UT, dims = 1:15)
#DimPlot(pbmc.UT, reduction = "umap")
#DimPlot(pbmc.UT, group.by = "cell_type", label = TRUE) + ggtitle("Clusters by Cell Types")
saveRDS(pbmc.UT, file = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/pbmc.UT.rds")
```
```{r}
pbmc.UT <- readRDS("/home/inf-21-2023/thesis/scRNAseq_data_oelen/pbmc.UT.rds")

Idents(pbmc.UT) <- factor(Idents(pbmc.UT), levels = sort(as.numeric(levels(Idents(pbmc.UT)))))
idents_sorted <- sort(unique(Idents(pbmc.UT)))

# Check if the number of unique identities matches the available age range entries
num_idents <- length(idents_sorted)
num_ages <- length(meta.data$age_range)

age_mapping <- setNames(meta.data$age_range[1:num_idents], idents_sorted)

# Assign age ranges to cells based on their identity
pbmc.UT@meta.data$age_range <- age_mapping[as.character(Idents(pbmc.UT))]

# Extract metadata from the pbmc.UT object

metadata_table <- pbmc.UT@meta.data

# Calculate total number of subjects
num_subjects <- length(unique(metadata_table$subject))

# Calculate total number of cells
total_cells <- nrow(metadata_table)

# Calculate sample size by cell type
cell_type_size <- as.data.frame(table(metadata_table$cell_type))

# Calculate sample size by age range
age_range_size <- as.data.frame(table(metadata_table$age_range))


# Add cell type and age range breakdowns
cell_type_breakdown <- paste(cell_type_size$Var1, ":", cell_type_size$Freq, collapse = ", ")
age_range_breakdown <- paste(age_range_size$Var1, ":", age_range_size$Freq, collapse = ", ")


summary_table <- data.frame(
  "Metric" = c("Total Cells", "Total Subjects", "Unique Cell Types", "Unique Age Ranges"),
  "Count" = c(total_cells, num_subjects, nrow(cell_type_size), nrow(age_range_size)),
  "Breakdown" = c(NA, NA, cell_type_breakdown, age_range_breakdown)
)

# Display the summary table
kable(summary_table, caption = "PBMC Metadata Summary", format = "html", table.attr = "class='table table-bordered'")

ggplot(metadata_table, aes(x = cell_type, fill = Sex)) +
  geom_bar(position = "dodge") +
  theme_minimal() +
  labs(title = "Cell Type Distribution by Sex", x = "Cell Type", y = "Count") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_fill_manual(values = c("Male" = "steelblue", "Female" = "pink"))

#saveRDS(pbmc.UT, file = "/home/inf-21-2023/thesis/scRNAseq_data_oelen/pbmc.UT.rds")
```
```{r}
pbmc.UT <- readRDS("/home/inf-21-2023/thesis/scRNAseq_data_oelen/pbmc.UT.rds")
DimPlot(pbmc.UT, group.by = "cell_type", label = TRUE)
```

```{r}
dittoBarPlot(pbmc.UT, var = "cell_type", group.by = "Sex", main = NULL, xlab = "Sex")
dittoBarPlot(pbmc.UT, var = "cell_type", group.by = "subject", main = "Cell type distribution between females and males", xlab = "Subject ID", ylab = NULL,split.by = "Sex", split.adjust = list(scales = "free_x"))

```
Cell frequencies
```{r}
set.seed(123)
# calculate cell type proportions
props <- getTransformedProps(clusters = pbmc.UT$cell_type,
                             sample = pbmc.UT$subject, transform = "logit")

# transform frequency into a dataframe
freq_df <- props$Proportions %>%
        as.data.frame() %>%
        tidyr::pivot_wider(names_from = "clusters",
                    values_from = "Freq",
                    ) %>%
        tibble::column_to_rownames("sample") %>%
        na.omit()

freq_matrix <- as.matrix(freq_df)

# Create the heatmap
Heatmap(freq_matrix,
        name = "Frequency",
       # col = colorRamp2(c(min(freq_matrix), max(freq_matrix)), c("yellow", "red")),
        show_row_names = TRUE,
        show_column_names = TRUE,
        cluster_columns = FALSE,
        row_names_gp = gpar(fontsize=8))
```
```{r}
#variance info
vari_df <- pbmc.UT@meta.data %>%
  distinct(subject, Sex) %>%
  tibble::remove_rownames() %>%
  tibble::column_to_rownames("subject")

#transform into dummy variable matrix
design <- model.matrix(~ 0 + Sex , data = vari_df)

```
```{r}
adonis <- vegan::adonis2(freq_df ~ Sex, data = vari_df, method = "euclidean", permutations = 999)

adonis <- adonis %>%
  filter(!grepl("Total", rownames(.))) %>%
  arrange(SumOfSqs)

vari_plot <- adonis %>%
  mutate(row_names = factor(rownames(adonis), levels = unique(rownames(adonis)))) %>%
  ggplot(aes(fill = row_names, y = R2, x = 1)) +
  geom_bar(stat="identity", position = "stack") +
  coord_flip() +
  theme_classic() +
  labs(x="", y="Variance explained", fill="Variables",
       title = "Proportion of cell type frequency variance among donors") +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())

vari_plot
```

```{r, eval = FALSE, echo = FALSE}

grep("^ESR",rownames(pbmc.UT@assays$RNA$counts),value = TRUE) # Estrogen receptor
grep("^AR$",rownames(pbmc.UT@assays$RNA$counts),value = TRUE) # $ defines exact match # Androgen receptor
grep("^PGR",rownames(pbmc.UT@assays$RNA$counts),value = TRUE) # Progesterone receptor
grep("^NR3C2",rownames(pbmc.UT@assays$RNA$counts),value = TRUE) # Mineralocorticoid receptor
grep("^GPER",rownames(pbmc.UT@assays$RNA$counts),value = TRUE)
```
```{r}

receptor_genes <- c("^ESR1$", "^ESR2$", "^NR3C1$", "^NR3C2$", "^PGR", "^AR$", "^GPER", "^LGR1", "^FSHR", "^LHR", "^LCGR",
"^LGR2", "^ULG5", "^GNRH", "^LHRH","^LHRHR","LSH-R","HHG", "^GRH$")

# Extract all genes from the Seurat object
all_genes <- rownames(pbmc.UT)

filtered_genes <- grep(paste(receptor_genes, collapse = "|"), all_genes, value = TRUE)

# Calculate average expression of receptor genes across cell types
#receptor_expression <- AggregateExpression(pbmc.UT, features = filtered_genes, return.seurat = TRUE)
#receptor_expression <- AggregateExpression(pbmc.UT, return.seurat = TRUE)
receptor_expression <- AverageExpression(pbmc.UT, return.seurat = TRUE, group.by = "cell_subtype")



cell_types_order <- c("B", "plasma B", "pDC", "mDC", "NKdim", "NKbright", "NK", "th1 CD4T", "th2 CD4T", "naive CD4T", "memory CD4T","reg CD4T", "naive CD8T ","memory CD8T","mono 1", "mono 2" ,"mono 4", "hemapoietic stem", "megakaryocyte", "unkown")
receptors_order <- c("ESR1", "ESR2", "NR3C1", "NR3C2", "PGRMC1", "PGRMC2", "AR", "GPER1", "GNRH1")


# Convert to factor with the specified order
pbmc.UT@meta.data$cell_subtype <- factor(pbmc.UT@meta.data$cell_subtype, levels = cell_types_order)

# Replace NA values with "unkown"
pbmc.UT@meta.data$cell_subtype[is.na(pbmc.UT@meta.data$cell_subtype)] <- "unkown"


#hh <- DoHeatmap(receptor_expression, features = filtered_genes, size = 2, draw.lines = FALSE)

  hh <- DoHeatmap(pbmc.UT,
                       features = receptors_order,
                       group.by = "cell_subtype") + theme(axis.text.y = element_text(size = 6),  # Adjust y-axis text size (genes)
                   axis.text.x = element_text(size = 5, angle = 45, hjust = 1))  # Adjust x-axis text size (cell types)
ggsave(hh, filename = "pbmc_exp.pdf")

```
```{r}

percent_stats <- Percent_Expressing(seurat_object = pbmc.UT, features = filtered_genes, threshold = 0, group_by = "subject")
plot <- DotPlot(object = pbmc.UT, features = filtered_genes, group.by = "cell_subtype") + coord_flip() + theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(plot, filename = "pbmc_percent_exp.pdf")
#plot_data <- plot$data %>%
#  select(pct.exp, id)

```
```{r}

cell_types_order <- c("B", "plasma B", "pDC", "mDC", "NKdim", "NKbright", "NK", "th1 CD4T", "th2 CD4T", "naive CD4T", "memory CD4T","reg CD4T", "naive CD8T ","memory CD8T","mono 1", "mono 2" ,"mono 4", "hemapoietic stem", "megakaryocyte", "unkown")
receptors_order <- c("ESR1", "ESR2", "NR3C1", "NR3C2", "PGRMC1", "PGRMC2", "AR", "GPER1", "GNRH1")

# Replace NA values with "unkown"
pbmc.UT@meta.data$cell_subtype[is.na(pbmc.UT@meta.data$cell_subtype)] <- "unkown"

# Convert to factor with the specified order
pbmc.UT@meta.data$cell_subtype <- factor(pbmc.UT@meta.data$cell_subtype, levels = cell_types_order)


plot_data <- DotPlot(pbmc.UT,
                     features = receptors_order,
                     group.by = "cell_subtype") +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(plot_data, filename = "pbmc_percent_exp.pdf")

```
Differential expression analysis
```{r}

```
```{r}
sessionInfo()
```
